Tests/Functional/Controller/VariantControllerTest.php

```diff
@@ -217,10 +430,150 @@ public function testDelete()
     private function getProductFactory()
     {
         return $this->getContainer()->get('sulu_product.product_factory');
     }
+
+    /**
+     * @return CurrencyRepository
+     */
+    private function getCurrencyRepository()
+    {
+        return $this->getContainer()->get('sulu_product.currency_repository');
+    }
+
+    /**
+     * Compares attributes data with response data.
+     *
+     * @param array $attributes
+     * @param array $response
+     */
+    private function checkAttributes(array $attributes, array $response)
+    {
+        // Check attributes.
+        $this->assertCount(count($attributes), $response['attributes']);
+
+        // Response order independent comparison of attributes:
+        // For all given attributes compare with response data.
+        foreach ($response['attributes'] as $responseAttribute) {
+            $match = null;
+            foreach ($attributes as $index => $attribute) {
+                if ($attribute['attributeId'] === $responseAttribute['attributeId']) {
+                    $match = $attribute;
+                    unset($attributes[$index]);
+                    break;
+                }
+            }
+            $this->assertNotNull($match);
+            $this->assertEquals($match['attributeValueName'], $responseAttribute['attributeValueName']);
+        }
+    }
+
+    /**
+     * Compares prices data with response data.
+     *
+     * @param array $prices
+     * @param array $response
+     */
+    private function checkPrices(array $prices, array $response)
+    {
+        // Check prices.
+        $this->assertCount(count($prices), $response['prices']);
+
+        foreach ($response['prices'] as $responsePrice) {
+            $match = null;
+            foreach ($prices as $index => $price) {
+                if ($price['currency']['id'] === $responsePrice['currency']['id']) {
+                    $match = $price;
+                    unset($prices[$index]);
+                    break;
+                }
+            }
+            $this->assertNotNull($match);
+            $this->assertEquals($match['price'], $responsePrice['price']);
+        }
+    }
+
+    /**
+     * Returns the product type of a variant.
+     *
+     * @return Type
+     */
+    private function getVariantTypeReference()
+    {
+        return $this->getEntityManager()->getReference(Type::class, $this->retrieveTypeIdByKey('PRODUCT_VARIANT'));
+    }
+
+    /**
+     * Maps product type string to its corresponding id.
+     *
+     * @param string $key
+     *
+     * @return int
+     */
+    private function retrieveTypeIdByKey($key)
+    {
+        $productTypesMap = $this->getContainer()->getParameter('sulu_product.product_types_map');
+
+        return $productTypesMap[$key];
+    }
+
+    /**
+     * @param array $validationResponse
+     * @param array $requiredAttributes
+     */
+    private function checkValidationErrorAttributes(array $validationResponse, array $requiredAttributes)
```

Comment:
use the phpunit-way to define such methods `assertValidationErrorAttributes(array $requiredAttributes, array$validationResponse)`